<launch>
  <!-- package configs -->
  <let name="project_name" value="d2_controller"/>
  <let name="project_namespace" value="d2::controller"/>
  <let name="project_share" value="$(find-pkg-share $(var project_name))"/>

  <!-- global args -->
  <arg name="namespace" default=""/>
  <arg name="container_name" default="d2_controller_container"/>
  <arg name="params_file" default="$(var project_share)/config/tc2025.param.yaml"/>
  <arg name="use_sim_time" default="false"/>

  <!-- launch args -->
  <arg name="use_pose_transformer" default="false"/>
  <arg name="use_global_plan_transformer" default="false"/>
  <arg name="controller_name" default="d2_controller"/>
  <arg name="pose_topic_name" default="$(var namespace)/pose"/>
  <arg name="global_plan_topic_name" default="$(var namespace)/waypoint_path_teb"/>
  <arg name="global_plan_without_avoidance_topic_name" default="$(var namespace)/waypoint_path_pp"/>
  <arg name="lidar_points_topic_name" default="$(var namespace)/livox/lidar_192_168_1_103"/>
  <arg name="vel_limit_topic_name" default="$(var namespace)/vel_limit"/>
  <arg name="accel_limit_topic_name" default="$(var namespace)/accel_limit"/>
  <arg name="cmd_vel_topic_name" default="$(var namespace)/cmd_vel"/>
  <arg name="obstacles_topic_name" default="$(var namespace)/obstacles"/>

  <let name="controller_pose_topic_name" value="pose/with_cov/transformed" if="$(var use_pose_transformer)"/>
  <let name="controller_pose_topic_name" value="$(var pose_topic_name)" unless="$(var use_pose_transformer)"/>
  <let name="controller_global_plan_topic_name" value="global_plan/transformed" if="$(var use_global_plan_transformer)"/>
  <let name="controller_global_plan_topic_name" value="$(var global_plan_topic_name)" unless="$(var use_global_plan_transformer)"/>
  <let name="controller_global_plan_without_avoidance_topic_name" value="global_plan_without_avoidance/transformed" if="$(var use_global_plan_transformer)"/>
  <let name="controller_global_plan_without_avoidance_topic_name" value="$(var global_plan_without_avoidance_topic_name)" unless="$(var use_global_plan_transformer)"/>
  <let name="tf_topic_name" value="$(var namespace)/tf"/>
  <let name="diagnostics_topic_name" value="$(var namespace)/diagnostics"/>

  <!-- namespace -->
  <push_ros_namespace namespace="$(var namespace)/$(var controller_name)"/>

  <!-- global parameters -->
  <set_parameter name="use_sim_time" value="$(var use_sim_time)"/>

  <!-- controller -->
  <node
      pkg="$(var project_name)"
      exec="local_planner_node"
      name="local_planner"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="pose/with_cov" to="$(var controller_pose_topic_name)"/>
    <remap from="global_plan" to="$(var controller_global_plan_without_avoidance_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
  <node
      pkg="$(var project_name)"
      exec="local_planner_node"
      name="local_planner_for_avoidance"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="pose/with_cov" to="$(var controller_pose_topic_name)"/>
    <remap from="global_plan" to="$(var controller_global_plan_topic_name)"/>
    <remap from="local_plan" to="local_plan_for_avoidance"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
  <node
      pkg="$(var project_name)"
      exec="obstacle_avoidance_planner_node"
      name="obstacle_avoidance_planner"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="plan" to="local_plan_for_avoidance"/>
    <remap from="avoidanced_plan" to="local_plan_avoidanced"/>
    <remap from="obstacles" to="/obstacles"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
  <!-- <node
      pkg="$(var project_name)"
      exec="local_planner_node"
      name="local_planner_for_avoidance"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="pose/with_cov" to="$(var controller_pose_topic_name)"/>
    <remap from="global_plan" to="$(var controller_global_plan_topic_name)"/>
    <remap from="local_plan" to="local_plan_for_avoidance"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node> -->
  <node
      pkg="$(var project_name)"
      exec="pure_pursuit_node"
      name="pure_pursuit"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <!-- <remap from="local_plan" to="local_plan_for_avoidance"/> -->
    <remap from="obstacles" to="$(var obstacles_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
    <remap from="cmd_vel" to="target_vel"/>
  </node>
  <node
      pkg="$(var project_name)"
      exec="pure_pursuit_node"
      name="pure_pursuit_for_avoidance"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <param name="lookahead_distance" value="3.0"/>
    <remap from="local_plan" to="local_plan_avoidanced"/>
    <remap from="obstacles" to="$(var obstacles_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
    <remap from="cmd_vel" to="target_vel"/>
  </node>
  <!-- <node
      pkg="$(var project_name)"
      exec="pure_pursuit_node"
      name="pure_pursuit_with_avoidance"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <param name="enable_avoidance" value="true"/>
    <remap from="obstacles" to="$(var obstacles_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
    <remap from="cmd_vel" to="target_vel"/>
  </node> -->
  <node
      pkg="$(var project_name)"
      exec="obstacle_vel_limitter_node"
      name="obstacle_vel_limitter"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="lidar/points" to="$(var lidar_points_topic_name)"/>
    <remap from="obstacle_vel_limit" to="vel_limit"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
  <node
      pkg="$(var project_name)"
      exec="vel_limiter_node"
      name="vel_limiter"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <param name="vel_limit.topics" value="[vel_limit, $(var vel_limit_topic_name)]"/>
    <remap from="cmd_vel" to="$(var cmd_vel_topic_name)"/>
    <!-- <remap from="vel_limit" to="$(var vel_limit_topic_name)"/> -->
    <remap from="accel_limit" to="$(var accel_limit_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
  
  <!-- geometry_transformer -->
  <node
      if="$(var use_pose_transformer)"
      pkg="geometry_transformer"
      exec="pose_with_covariance_transformer_node"
      name="pose_with_covariance_transformer"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="pose/with_cov" to="$(var pose_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
  <node
      if="$(var use_global_plan_transformer)"
      pkg="geometry_transformer"
      exec="path_transformer_node"
      name="global_plan_transformer"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="path" to="$(var global_plan_topic_name)"/>
    <remap from="path/transformed" to="global_plan/transformed"/>
    <remap from="/tf" to="$(var tf_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
  <node
      if="$(var use_global_plan_transformer)"
      pkg="geometry_transformer"
      exec="path_transformer_node"
      name="global_plan_without_avoidance_transformer"
      output="screen">
    <param from="$(var params_file)" allow_substs="true"/>
    <remap from="path" to="$(var global_plan_without_avoidance_topic_name)"/>
    <remap from="path/transformed" to="global_plan_without_avoidance/transformed"/>
    <remap from="/tf" to="$(var tf_topic_name)"/>
    <remap from="/diagnostics" to="$(var diagnostics_topic_name)"/>
  </node>
</launch>
